samples/storage/storage-postgres/src/postgres_wallet.rs

```diff
@@ -0,0 +1,1781 @@
+extern crate libc;
+extern crate time;
+extern crate indy_common;
+extern crate indy_crypto;
+extern crate serde_json;
+
+use indy_common::api::ErrorCode;
+use indy_common::utils::sequence::SequenceUtils;
+use indy_common::utils::cstring::CStringUtils;
+use indy_common::utils::crypto::base64;
+use indy_common::wallet_storage::storage::{WalletStorage, StorageRecord, StorageIterator, Tag, TagName, EncryptedValue};
+use postgres_storage::storage::WalletStorageType;
+use indy_common::wallet_storage::language;
+use indy_common::errors::wallet::WalletStorageError;
+
+use self::libc::c_char;
+
+use std::collections::HashMap;
+use std::ffi::CString;
+use std::sync::Mutex;
+use std::str;
+
+
+struct PostgresStorageContext {
+    // TODO save handle, config and credentials in case we need to re-connect to database
+    _xhandle: i32,        // reference returned to client to track open wallet connection
+    id: String,          // wallet name
+    _config: String,      // wallet config
+    _credentials: String, // wallet credentials
+    phandle: Box<::postgres_storage::PostgresStorage>  // reference to a postgres database connection
+}
+
+#[derive(Debug, Clone)]
+struct PostgresWalletRecord {
+    id: CString,
+    type_: CString,
+    value: Vec<u8>,
+    tags: CString
+}
+
+#[derive(Debug, Clone)]
+struct PostgresWalletRecordSet {
+    idx: usize,
+    records: Vec<PostgresWalletRecord>,
+    count: usize
+}
+
+lazy_static! {
+    // store a PostgresStorage object (contains a connection) 
+    static ref POSTGRES_OPEN_WALLETS: Mutex<HashMap<i32, PostgresStorageContext>> = Default::default();
+}
+
+lazy_static! {
+    // metadata for active wallets
+    static ref POSTGRES_ACTIVE_METADATAS: Mutex<HashMap<i32, CString>> = Default::default();
+}
+
+lazy_static! {
+    // cache of Postgres fetched records
+    static ref POSTGRES_ACTIVE_RECORDS: Mutex<HashMap<i32, PostgresWalletRecord>> = Default::default();
+}
+
+lazy_static! {
+    // cache of active Postgres searches
+    // TODO figure out  athread-safe PostgresStorageIterator
+    // static ref POSTGRES_ACTIVE_SEARCHES: Mutex<HashMap<i32, Box<::postgres_storage::PostgresStorageIterator>>> = Default::default();
+    static ref POSTGRES_ACTIVE_SEARCHES: Mutex<HashMap<i32, PostgresWalletRecordSet>> = Default::default();
+}
+
+pub struct PostgresWallet {}
+
+impl PostgresWallet {
```

Comment:
I suggest don't use structs to group functions. Modules are designed for this in Rust. There is some old examples of this approach in our codebase, but now we are working on removing this.